source/common/ssl/context_manager_impl.h

```diff
@@ -27,20 +28,30 @@ class ContextManagerImpl final : public ContextManager {
    * admin purposes. When a caller frees a context it will tell us to release it also from the list
    * of contexts.
    */
-  void releaseContext(Context* context);
+  void releaseClientContext(ClientContext* context);
+  void releaseServerContext(ServerContext* context, const std::string& listener_name,
+                            const std::vector<std::string>& server_names);
 
   // Ssl::ContextManager
   Ssl::ClientContextPtr createSslClientContext(Stats::Scope& scope,
                                                ClientContextConfig& config) override;
-  Ssl::ServerContextPtr createSslServerContext(Stats::Scope& scope,
-                                               ServerContextConfig& config) override;
-  size_t daysUntilFirstCertExpires() override;
-  void iterateContexts(std::function<void(Context&)> callback) override;
+  Ssl::ServerContextPtr createSslServerContext(const std::string& listener_name,
+                                               const std::vector<std::string>& server_names,
+                                               Stats::Scope& scope, ServerContextConfig& config,
+                                               bool skip_context_update) override;
+  Ssl::ServerContext* findSslServerContext(const std::string& listener_name,
+                                           const std::string& server_name) const override;
+  size_t daysUntilFirstCertExpires() const override;
+  void iterateContexts(std::function<void(const Context&)> callback) override;
 
 private:
   Runtime::Loader& runtime_;
   std::list<Context*> contexts_;
-  std::mutex contexts_lock_;
+  mutable std::shared_timed_mutex contexts_lock_;
+  std::unordered_map<std::string, std::unordered_map<std::string, ServerContext*>> map_exact_;
+  std::unordered_map<std::string, std::unordered_map<std::string, ServerContext*>> map_wildcard_;
+
+  static bool isWildcardServerName(const std::string& name);
```

Comment:
nit: functions go above variables (move to line 48 followed by newline)

Reply:
Done.